### PR TITLE
WireRec for scala2

### PR DESCRIPTION
**Pull request by mbore**

### CODE REVIEW

File: macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala [hunk 1]
```diff
@@ -32,12 +32,12 @@ private[macwire] class CompanionCrimper [C <: blackbox.Context, T: C#WeakTypeTag
 
   lazy val applyParamLists: Option[List[List[Symbol]]] = apply.map(_.asMethod.paramLists)
 
-  def wireParams(paramLists: List[List[Symbol]]): List[List[Tree]] = paramLists.map(_.map(p => dependencyResolver.resolve(p, p.typeSignature)))
+  def wireParams(dependencyResolver: DependencyResolverType)(paramLists: List[List[Symbol]]): List[List[Tree]] = paramLists.map(_.map(p => dependencyResolver.resolve(p, p.typeSignature)))
```
Comment (adamw) [inline macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala:35]: can't we have `dependencyResolver` as a parameter of `CompanionCrimper`?

File: macros/src/main/scala-2/com/softwaremill/macwire/internals/ConstructorCrimper.scala [hunk 2]
```diff
@@ -44,22 +44,22 @@ private[macwire] class ConstructorCrimper[C <: blackbox.Context, T: C#WeakTypeTa
 
   lazy val constructorParamLists: Option[List[List[Symbol]]] = constructor.map(_.asMethod.paramLists.filterNot(_.headOption.exists(_.isImplicit)))
 
-  lazy val constructorArgs: Option[List[List[Tree]]] = log.withBlock("Looking for targetConstructor arguments") {
```
Comment (adamw) [inline macros/src/main/scala-2/com/softwaremill/macwire/internals/ConstructorCrimper.scala:47]: same here, maybe we can introduce a second arg list to `ConstructorCrimper` and move the dependency there

File: macros/src/main/scala-2/com/softwaremill/macwire/internals/DependencyResolver.scala [hunk 3]
```diff
@@ -4,7 +4,7 @@ import com.softwaremill.macwire.internals.EligibleValuesFinder.Scope.LocalForwar
 
 import scala.reflect.macros.blackbox
 
-private[macwire] class DependencyResolver[C <: blackbox.Context](val c: C, debug: Logger) {
+private[macwire] class DependencyResolver[C <: blackbox.Context, TypeC <: C#Type, TreeC <: C#Tree](val c: C, debug: Logger)(resolutionFallback: TypeC => TreeC) {
```
Comment (adamw) [inline macros/src/main/scala-2/com/softwaremill/macwire/internals/DependencyResolver.scala:7]: I think we should be able to use `c`-dependent types in the second argument list

**Comment by adamw** [comment]
Seems like some method calls in the akka module need fixing

File: macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala (hunk 1 above)
Comment (mbore) [reply macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala:35]: I tried to define CompanioCrimper like this:
```Scala
private[macwire] class CompanionCrimper [C <: blackbox.Context, T: C#WeakTypeTag](val c: C, log: Logger)(dependencyResolver: DependencyResolver[c.type, c.universe.Type, c.universe.Tree]) {
...
}
```

but it fails with 
```
[error][...]/macwire/macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala:5:145: not found: value c
[error] private[macwire] class CompanionCrimper [C <: blackbox.Context, T: C#WeakTypeTag](val c: C, log: Logger)(dependencyResolver: DependencyResolver[c.type, c.universe.Type, c.universe.Tree]) {
[error]                                                                                                                                                 ^
[error][...]/macwire/macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala:5:153: not found: value c
[error] private[macwire] class CompanionCrimper [C <: blackbox.Context, T: C#WeakTypeTag](val c: C, log: Logger)(dependencyResolver: DependencyResolver[c.type, c.universe.Type, c.universe.Tree]) {
[error]                                                                                                                                                         ^
[error][...]/macwire/macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala:5:170: not found: value c
[error] private[macwire] class CompanionCrimper [C <: blackbox.Context, T: C#WeakTypeTag](val c: C, log: Logger)(dependencyResolver: DependencyResolver[c.type, c.universe.Type, c.universe.Tree]) {
```

Comment (adamw) [reply macros/src/main/scala-2/com/softwaremill/macwire/internals/CompanionCrimper.scala:35]: ah right, you can't use dependent types in constructors, forgot about that :)

**Pull request merged**